### PR TITLE
Help wanted: flattening negated nested divisions

### PR DESCRIPTION
**Pull request by niklasdewally**
The purpose of this PR is to get feedback on my ongoing work on flattening nested divs and to debug a failing testcase.

The test case in question is `div_undefzero-05-nested-noteq`, which is defined as:

```
find x : int(5..20)
find y : int(0..5)
find z : int(0..6)

such that !(x / (y/z) = 10)
```

This PR is completed in #454

### CODE REVIEW

**Comment by niklasdewally** [comment]
@ozgurakgun what are the undefinedness semantics of (a/y !=z)?

The existing `div_to_bubble` rule seems to be trying to do 

Not(Neq(UnsafeDiv(x,y),z)) ~> Not(And[Neq(SafeDiv(x,y)z), Neq(y,0)]) ~> And[Eq(SafeDiv(x,y),z),Eq(y,0)]

which is wrong.


This is causing `div_undefzero-05-nested-noteq` to fail.

**Comment by niklasdewally** [comment]
I have made `to_aux_var` fail when it has an undefined thing inside it so that we don't bring something to the top that should've been bubbled; but this doesn't make a difference...

**Comment by niklasdewally** [comment]
`div-05`, which does (a!= b/c), works as expected, so it might be more to do with double negations? 

I know that its probably not my flattening rules doing it, so I'm a bit stuck...

**Comment by niklasdewally** [comment]
```
Rule applicable: negated_neq_to_eq ([("Base", 8800)]), to expression: Not((SafeDiv(y, z) != 0)), resulting in: (SafeDiv(y, z) = 0) new_top=And([])
```

This is weird....

**Comment by niklasdewally** [comment]
bubbling up the auxvar instead of posting it to the top level is also very wrong...

**Comment by niklasdewally** [comment]
Heres the rule trace: 
https://gist.github.com/niklasdewally/39a3a938e477f4b109202850e33e3b66

**Comment by ozgurakgun** [comment]
```
(a/y !=z)

~~> parsed as:
Neq(Div(a,y), z)

~~> make safe
Eq(Bubble(SafeDiv(a,y), Neq(y,0)), z)

~~> bubble-up
Bubble(Eq(SafeDiv(a,y), z), Neq(y, 0))

~~> bubble-to-and since first arg is bool
And(Eq(SafeDiv(a,y), z), Neq(y, 0))
```

**Comment by ozgurakgun** [comment]
the nested case is interesting

```
!(x/(y/z) = 10)

~~> parsed as
Not(Eq(Div(x, Div(y,z)), 10))

~~> make-safe the inside div
Not(Eq(Div(x, Bubble(SafeDiv(y,z), Neq(z,0)), 10))

~~> (*)
Not(Eq(SafeDiv(x, Bubble(SafeDiv(y,z), Neq(z,0)), Neq(Bubble(SafeDiv(y,z), Neq(z,0), 0), 10))

```

(*) if we allow make-safe the outside div without bubbling up first, we would get. not wrong, but if we bubbled-up first and 
then converted the outside Div to SafeDiv we probably won't repeat work?

So rule_Div_to_SafeDiv should refuse to apply i fthe second arg is a Bubble.

Happy for you to try both options (both should be correct) and see the rule application trace before we commit to one.

**Comment by niklasdewally** [comment]
> ```
> (a/y !=z)
> 
> ~~> parsed as:
> Neq(Div(a,y), z)
> 
> ~~> make safe
> Eq(Bubble(SafeDiv(a,y), Neq(y,0)), z)
> 
> ~~> bubble-up
> Bubble(Eq(SafeDiv(a,y), z), Neq(y, 0))
> 
> ~~> bubble-to-and since first arg is bool
> And(Eq(SafeDiv(a,y), z), Neq(y, 0))
> ```

On further testing, I think this bit works well; I suspect a missing "if undefined return NA" case somewhere....

My current stumbling block is that the first one below works, but the second doesn't:
```
x/(y/z) != 10
```

```
!(x/(y/z) = 10) 
```

**Comment by ozgurakgun** [comment]
here

```
x/(y/z) != 10

!(x/(y/z) = 10) 
```

both !=0 bubbles should stay in the !() for the second one, end up at the top level for the first one.

Did you test `x/y != 10` and `!(x/y=10)` as well? Do they work as expected?

**Comment by niklasdewally** [comment]
> here
> 
> ```
> x/(y/z) != 10
> 
> !(x/(y/z) = 10) 
> ```
> 
> both !=0 bubbles should stay in the !() for the second one, end up at the top level for the first one.
> 
> Did you test `x/y != 10` and `!(x/y=10)` as well? Do they work as expected?

We already had a test case for the first; just tried the second and that passes too.

**Comment by niklasdewally** [comment]
> (*) if we allow make-safe the outside div without bubbling up first, we would get. not wrong, but if we bubbled-up first and then converted the outside Div to SafeDiv we probably won't repeat work?
> 
> So rule_Div_to_SafeDiv should refuse to apply i fthe second arg is a Bubble.

Do you mean that we should not create a bubble if we have a bubble as a child?

I tried bottom up vs top-down bubbling previously, and that didn't have an affect on the result, although not with the above condition.

**Comment by ozgurakgun** [comment]
as rule traces if you could produce something of the style I used above that would be the easiest for me to follow/debug

something like

```
expr1

~~> rulename
expr2

~~> rulename
expr3
```

**Comment by ozgurakgun** [comment]
I think both (created nested bubbles and waiting untill they bubble up) should work correctly actually, but one will be fewer steps - I suspect the latter will be fewer steps.

**Comment by niklasdewally** [comment]
I'm not sure if this is what you meant, but it still doesn't work.

**Comment by niklasdewally** [comment]
Makes smaller models though :)

**Comment by niklasdewally** [comment]
This made the rule ordering a lot shorter... By process of elimination, could we be missing something in distribute_not_over_and? This runs after each bubble while there are still some UnsafeDivs left in the tree. not(eq) to neq , flattening etc wait until we only have SafeDivs

File: crates/conjure_core/src/rules/bubble.rs [hunk 1]
```diff
@@ -75,6 +75,17 @@ fn div_to_bubble(expr: &Expression, _: &Model) -> ApplicationResult {
         return Err(RuleNotApplicable);
     }
     if let Expression::UnsafeDiv(_, a, b) = expr {
+        // bubble bottom up
+        if a.can_be_undefined() || b.can_be_undefined() {
+            return Err(RuleNotApplicable);
+        }
+
+        // only do one bubble at a time
+        if matches!(**a, Expression::Bubble(_, _, _)) || matches!(**b, Expression::Bubble(_, _, _))
```
Comment (ozgurakgun) [inline crates/conjure_core/src/rules/bubble.rs:84]: we should be able to do multiple bubbles at once too.

```
{a @ bubble_a} + {b @ bubble_b}

~~> bubble-up
{a + b @ bubble_a /\ bubble_b}
```

is a valid transformation. 

I know this rule doesn't do `+` but I noticed the check here, so noting here.

Comment (ozgurakgun) [reply crates/conjure_core/src/rules/bubble.rs:84]: in fact the `bubble_up` rule already does this. ignore.

**Comment by lixitrixi** [comment]
I remember having an issue with `!= ~> !(=)`. Are the semantics we use in that case correct? This is for the final step after bubbling etc.

**Comment by niklasdewally** [comment]
That means that, unfor

> I remember having an issue with `!= ~> !(=)`. Are the semantics we use in that case correct? This is for the final step after bubbling etc.

We do this when everything is fully defined. This seems to work, as the non-nested !(=) and != tests work.

**Comment by niklasdewally** [comment]
@lixitrixi and I wrote this on paper and theres nothing in the rewritten model that says that z cannot equal zero...

Yet, those are the solutions that we are missing?

**Comment by niklasdewally** [comment]
Infact, it says z=0!

**Comment by niklasdewally** [comment]
conjure gives us the following minion, which seems the same as what we get as output:

Screenshot 2024-11-15 at 22 07 36

**Comment by niklasdewally** [comment]
Its a domain evaluation issue!!!!

In our model,`__1` has domain {1..20}, but Conjure gives it domain {0..20}!

**Comment by ozgurakgun** [comment]
Yep, it needs to have 0 in its domain. Good debug!

**Comment by niklasdewally** [comment]
@ozgurakgun if this is all good,  I will write a commit message and clean this up to merge tomorrow.

File: crates/conjure_core/src/rules/utils.rs [hunk 2]
```diff
@@ -36,6 +37,33 @@ pub fn is_all_constant(expression: &Expr) -> bool {
...
+        [a] if a.return_type() == Some(ReturnType::Bool) => Some(a.clone()),
+        [_, _, ..] => {
+            for expr in exprs {
+                if expr.return_type() != Some(ReturnType::Bool) {
+                    return None;
+                }
+            }
+            Some(Expr::And(Metadata::new(), exprs.clone()))
+        }
+        _ => None,
```
Comment (ozgurakgun) [inline crates/conjure_core/src/rules/utils.rs:63]: this case should be redundant? why not change line 55 to be _ and remove line 63?

**Review by ozgurakgun — APPROVED** [review]
just a minor stylistic comment, looks good otherwise.

**Comment by ozgurakgun** [comment]
oh also there are a few extra files (2, a*txt) but I assume you know about these.

**Comment by niklasdewally** [comment]
Completed by #454

**Pull request closed without merging**